cmd/service-controller/definition_monitor_test.go

```diff
@@ -1133,6 +1133,63 @@ func TestUpdateAnnotatedServiceDefinition(t *testing.T) {
 				},
 			},
 		},
+		{
+			name: "six",
+			actual: types.ServiceInterface{
+				Address:  "six",
+				Protocol: "tcp",
+				Origin:   "annotation",
+				Ports:    []int{8080, 8081, 8082},
+				Targets: []types.ServiceInterfaceTarget{
+					{
+						Name:     "t1",
+						Selector: "foo=bar",
+						TargetPorts: map[int]int{
+							8080: 8080,
+							8081: 8081,
+							8082: 8082,
+						},
+					},
+				},
+			},
+			desired: types.ServiceInterface{
+				Address:  "five",
+				Protocol: "tcp",
+				Ports:    []int{8080, 8082, 8081},
```

Comment:
Try applying this patch to the `definition_monitor_test.go` and you will see the same behavior
reported through #865.

```
diff --git a/cmd/service-controller/definition_monitor_test.go b/cmd/service-controller/definition_monitor_test.go
index ca256ef7..f003a148 100644
--- a/cmd/service-controller/definition_monitor_test.go
+++ b/cmd/service-controller/definition_monitor_test.go
@@ -1155,7 +1155,7 @@ func TestUpdateAnnotatedServiceDefinition(t *testing.T) {
 			desired: types.ServiceInterface{
 				Address:  "five",
 				Protocol: "tcp",
-				Ports:    []int{8080, 8081, 8082},
+				Ports:    []int{8080, 8081, 8082, 8083},
 				Targets: []types.ServiceInterfaceTarget{
 					{
 						Name:     "t2",
@@ -1164,6 +1164,7 @@ func TestUpdateAnnotatedServiceDefinition(t *testing.T) {
 							8080: 8080,
 							8081: 8081,
 							8082: 8082,
+							8083: 8083,
 						},
 					},
 				},
@@ -1186,6 +1187,7 @@ func TestUpdateAnnotatedServiceDefinition(t *testing.T) {
 						8080: 8080,
 						8081: 8081,
 						8082: 8082,
+						8083: 8083,
 					},
 				},
 			},
```

Reply:
If the ports are actually different, then the two definitions are not compatible and the later one is used as a replacement. If they are compatible, then the targets are merged.

So in the case where you have four ports defined instead of three, then the expected behaviour is to take the second definition as a replacement for the first i.e. the first target will be lost as it is based on an definition deemed incompatible. Likewise if the protocol or some other aspect of the service is different. Only the targets for compatible definitions can be merged.

Reply:
That makes sense! Thank you.